### DISCUSSION TITLE
proposal: rename plugins to components with sub-categories of connectors, tools, and resources

### DISCUSSION THREAD

**Discussion by tayloramurphy** [meltano/meltano#6561] [General]
Previously discussed in https://github.com/meltano/internal-general/issues/333 (private repo).

Proposal is to have a broad renaming of `plugins` --> `components`. A sub-category from there could potentially be `connectors`, `tools` and `resources`.

@aaronsteers proposed a few "frames" which we can use to evaluate how we feel about this change.

- How we bucket or categorize plugins on hub.meltano.com.
- What appears in yaml to replace `plugins:` entry. (Another top-level entry or the new subcategories.)
- What do we call the upcoming SDK for plugins.
- What replaces `--plugin-type` and in the CLI?
- How does the wording work when we talk about plugins inheriting from other plugins.

**Comment by tayloramurphy** [comment]
> How we bucket or categorize plugins on hub.meltano.com.

Components would essentially be a drop-in replacement for plugins. Airflow, dbt, etc are components of your data stack.

> What appears in yaml to replace plugins: entry. (Another top-level entry or the new subcategories.)

I'm partial to dropping the top-level `plugins` key in favor of `connectors` (perhaps with `extractors` and `loaders` subkeys? or could those be new keys on the connector parent?) and `tools`. 

> SDK for Plugins

Maybe Component Extension SDK? Something along those lines. Not sold on Extension fully though

> What replaces --plugin-type and in the CLI?

`--component-type` seems fine. 

> How does the wording work when we talk about plugins inheriting from other plugins.

Funny enough, we already use components in https://docs.meltano.com/concepts/plugins `Meltano provides the glue to make these components work together smoothly and enables consistent [configuration](https://docs.meltano.com/guide/configuration) and [deployment](https://docs.meltano.com/guide/production).`

I don't think it's too much to go from this:

>If you’d like to use the same package (base plugin) in your project multiple times with [different configurations](https://docs.meltano.com/guide/configuration#multiple-plugin-configurations), you can add a new plugin that inherits from an existing one.

> The new plugin will inherit its parent’s [base plugin description](https://docs.meltano.com/concepts/plugins#project-plugins) and [configuration](https://docs.meltano.com/guide/configuration) as if they were defaults, which can then be overridden as appropriate.

> For performance reasons, inherited plugins with an identical pip_url to their parent share the parents underlying python virtualenv. If you would prefer to create a separate virtualenv for an inherited plugin, modify it’s pip_url to be different to its parent.

to this 

>If you’d like to use the same package (base component) in your project multiple times with [different configurations](https://docs.meltano.com/guide/configuration#multiple-plugin-configurations), you can add a new component that inherits from an existing one.

> The new component will inherit its parent’s [base component description](https://docs.meltano.com/concepts/plugins#project-plugins) and [configuration](https://docs.meltano.com/guide/configuration) as if they were defaults, which can then be overridden as appropriate.

> For performance reasons, inherited components with an identical pip_url to their parent share the parents underlying python virtualenv. If you would prefer to create a separate virtualenv for an inherited component, modify it’s pip_url to be different to its parent.

**Comment by pnadolny13** [comment]
@tayloramurphy I talked about this in office hours but wanted to note it here as well. I understand why we would want to bucket them for things like the hub, marketing, and generally for making it cleaner to talk about them. As a user though, having to re-organize my project and configurations for that reason is pretty annoying.

When it comes to the actual meltano.yml and organizing a project I like the idea of calling them all one thing. As a user it doesnt matter to me if something is a service/tool/resources/connector/component/plugin/etc. I could see them all being under a single top level key, or no key at all. Then on the hub we can group them however we'd like and easily regroup them over time.

If we are getting value out of bucketing them, for example if the backend uses its type as a way of determining what functionality to provide (like we do today with EL vs utility plugins) then I wonder if we could add something like `component_type` as an attribute to the plugin instead. That gives us a way to insulate the user from us changing names. Like:

```
  - name: tap-gitlab
    component_type: connector
    variant: meltanolabs
    pip_url: git+https://github.com/pnadolny13/tap-gitlab.git@author_usernames
    select:
    - projects.*
    - merge_requests.*
    - issues.*
```

In 6 months we can decide to call them integrations but we dont require every user to re-organize their project and configurations again to support our new names, they just update the component type or leave it and we have backwards compatibility. It's sort of like the pattern we've been using with capabilities, just list what it can do, for a tap meltano should manage catalogs and state but for superset it shouldnt.

Also I do have the control to organize them however I'd like using the [include_paths](https://github.com/meltano/squared/blob/a3c5e6c9d5f09c9b4c12db928745f6d2c6821a7a/data/meltano.yml#L5) so when it comes to the negative effects of a user having a giant blob of plugins/components sitting together, they can split them up like the terraform pattern.

What do you think? I know its more of a marketing concern but I wanted to understand how it affects users and projects.

**Reply by tayloramurphy** [reply]
> As a user it doesnt matter to me if something is a service/tool/resources/connector/component/plugin/etc

@pnadolny13 agreed and that's part of the motivation for this a bit. Reminds me a bit of the conversation I had with Edgar a while ago about more granular capabilities https://gitlab.com/meltano/hub/-/issues/187#note_815498390 It's possible we wouldn't even have to have the `type` declaration. 

> In 6 months we can decide to call them integrations but we dont require every user to re-organize their project and configurations again to support our new names, they just update the component type or leave it and we have backwards compatibility.

Agreed and I want to avoid that as well. We do have versioning in the meltano.yml file which could be a controller around this if we wanted to go down that path. 

I think the smallest increment on this is as primarily a marketing effort with possibly allowing `components` in place of `plugins` in the meltano.yml.

**Comment by tayloramurphy** [comment]
@aaronsteers wanted to ping you on this since you mentioned the tools vs utilities distinction. 

In poking around google and other sites, it seems like "tools" is a pretty accepted term for the different parts of a data stack: https://duckduckgo.com/?t=ffab&q=modern+data+stack&ia=web 

I like the framing of components as the high-level with the 2 main distinctions being connectors and tools.

**Reply by aaronsteers** [reply]
This one got lost in a GitHub notifications flood - circling back...

I do lean slightly towards 'connectors' (extractors, loaders, and mappers) and 'utilities' (everything else that's runnable or has a 'commands' interface) but perhaps I'm biased towards language that's more familiar to me after having used `Utilities` for a while now. 🤷 

Perhaps another take is that things like 'dbt' and 'sqlfluff' are definitely 'tools' but I don't immediately associate the word tools with other parts of the platform (especially "services") like 'Airflow', 'Superset', 'Postgres', and 'SSH' (a random bespoke example). My intuition (which might be wrong) feels like they all resonate as 'utilities' but maybe not as 'tools' in the sense that I normally think of them. 🤔 

Admittedly, this is very subjective. So, totally not going to say 'no' to "tools", but perhaps still worth more discussion.
